Fix aten.copy device mismatch bug in FakeTensor

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102664

Fixes `pytest ./generated/test_yizhou_wang_RODNet.py -k test_000` failure in https://github.com/pytorch/pytorch/issues/92670.

FakeTensor would raise an error upon trying to run `aten.copy` with inputs with different devices, although this is allowed behavior.

Also fix `aten.slice_scatter`, since it also takes args with different devices.

williamwen42 committed:
Fix aten.copy device mismatch bug in FakeTensor

[ghstack-poisoned]

williamwen42 committed:
Update on "Fix aten.copy device mismatch bug in FakeTensor"


Fixes `pytest ./generated/test_yizhou_wang_RODNet.py -k test_000` failure in https://github.com/pytorch/pytorch/issues/92670.

FakeTensor would raise an error upon trying to run `aten.copy` with inputs with different devices, although this is allowed behavior.


[ghstack-poisoned]

williamwen42 committed:
Update on "Fix aten.copy device mismatch bug in FakeTensor"


Fixes `pytest ./generated/test_yizhou_wang_RODNet.py -k test_000` failure in https://github.com/pytorch/pytorch/issues/92670.

FakeTensor would raise an error upon trying to run `aten.copy` with inputs with different devices, although this is allowed behavior.


[ghstack-poisoned]

yanboliang commented:
Can you coordinate with @davidberard98 to add other ops that support different devices? Thanks!

davidberard98 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -576,6 +576,24 @@ def index_put_(fake_mode, func, *args, **kwargs):
+# copy.default/copy.out is the same as index_put/index_put_
+@register_op_impl(aten.copy.default)
+def copy_default(fake_mode, func, *args, **kwargs):
+    return run_and_return_new_tensor_of_input_device(fake_mode, func, args, kwargs)
```
if you can add slice_scatter here, I think it should also fix test_NKI_AI_direct.py

williamwen42 replied:
done

williamwen42 committed:
Update on "Fix aten.copy device mismatch bug in FakeTensor"


Fixes `pytest ./generated/test_yizhou_wang_RODNet.py -k test_000` failure in https://github.com/pytorch/pytorch/issues/92670.

FakeTensor would raise an error upon trying to run `aten.copy` with inputs with different devices, although this is allowed behavior.


[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge

williamwen42 commented:
@pytorchbot merge

ezyang commented on test/test_fake_tensor.py:
```diff
@@ -582,6 +582,56 @@ def forward(self, input):
+            j1 = torch.tensor([2], device="cuda")
+            j2 = torch.tensor([2], device="cpu")
+            r3 = torch.ops.aten.index_put.default(x1, j1, y1)
+            r4 = torch.ops.aten.index_put.default(x2, j2, y2)
```
This code doesn't actually pass on eager though?

```
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (fc2aa23c)]$ python test.py
Traceback (most recent call last):
  File "/data/users/ezyang/b/pytorch/test.py", line 7, in <module>
    r1 = torch.ops.aten.index(x1, i1)
  File "/data/users/ezyang/b/pytorch/torch/_ops.py", line 1018, in __call__
    return self_._op(*args, **(kwargs or {}))
RuntimeError: indices should be either on cpu or on the same device as the indexed tensor (cpu)
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (fc2aa23c)]$ cat test.py
import torch

x1 = torch.rand(4, 4, device="cpu")
x2 = torch.rand(4, 4, device="cuda")
i1 = torch.tensor([0, 1], device="cuda")
i2 = torch.tensor([0, 1], device="cpu")
r1 = torch.ops.aten.index(x1, i1)
r2 = torch.ops.aten.index(x2, i2)

y1 = torch.rand(4, device="cpu")
y2 = torch.rand(4, device="cuda")
j1 = torch.tensor([2], device="cuda")
j2 = torch.tensor([2], device="cpu")
r3 = torch.ops.aten.index_put.default(x1, j1, y1)
r4 = torch.ops.aten.index_put.default(x2, j2, y2)
```
